File: medplum-java/medplum-model/src/main/java/com/fhir/Specimendefinition_typetestedPreference.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class Specimendefinition_typetestedPreference {
  public static final Specimendefinition_typetestedPreference PREFERRED = new Specimendefinition_typetestedPreference(Value.PREFERRED, "PREFERRED");

  public static final Specimendefinition_typetestedPreference ALTERNATE = new Specimendefinition_typetestedPreference(Value.ALTERNATE, "ALTERNATE");

  private final Value value;

  private final String string;

  Specimendefinition_typetestedPreference(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof Specimendefinition_typetestedPreference && this.string.equals(((Specimendefinition_typetestedPreference) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case PREFERRED:
        return visitor.visitPREFERRED();
      case ALTERNATE:
        return visitor.visitALTERNATE();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static Specimendefinition_typetestedPreference valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "PREFERRED":
        return PREFERRED;
      case "ALTERNATE":
        return ALTERNATE;
      default:
        return new Specimendefinition_typetestedPreference(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    PREFERRED,

    ALTERNATE,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitPREFERRED();

    T visitALTERNATE();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/Specimendefinition_typetestedPreference.ts
export type Specimendefinition_typetestedPreference =
    | Specimendefinition_typetestedPreference.Preferred
    | Specimendefinition_typetestedPreference.Alternate;

export const Specimendefinition_typetestedPreference = {
    Preferred: "preferred" as Specimendefinition_typetestedPreference.Preferred,
    Alternate: "alternate" as Specimendefinition_typetestedPreference.Alternate,

    _visit: <Result>(value: Specimendefinition_typetestedPreference, visitor: Specimendefinition_typetestedPreference._Visitor<Result>): Result => {
        switch (value) {
            case Specimendefinition_typetestedPreference.Preferred: return visitor.preferred();
            case Specimendefinition_typetestedPreference.Alternate: return visitor.alternate();
            default: return visitor._unknown();
        }
    },
};

export declare namespace Specimendefinition_typetestedPreference {
    export type Preferred = "preferred" & {
        __Preferred: void,
    };
    export type Alternate = "alternate" & {
        __Alternate: void,
    };

    export interface _Visitor<Result> {
        preferred: () => Result;
        alternate: () => Result;
        _unknown: () => Result;
    }
}
